vc4: Add zpos support

6by9 opened the pull request:
Adds support for zpos to the KMS driver.
The frameworks didn't apply the stored default value to the planes when reset, hence a framework patch first.

@mripard Does this look like an acceptable solution that can be upstreamed?

6by9 committed:
drm/vc4: Support zpos on all planes

Adds the zpos property to all planes, and creates the dlist
by placing the fragments in the correct order based on zpos.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

mripard commented:
Yes, it does :)

I'll rework it a bit to split it into several patches, but it'll work.

6by9 commented:
OK, great. I'll leave it with you to split as required, and we'll backport it when you've posted that.

I didn't go through all the drivers we'd looked at to undo the setting of default values in their plane state_reset functions, but I guess that could follow on too.

6by9 committed:
drm/object: Add drm_object_property_get_default_value() function

Upstream commit adf47b75297ebc71c53b6dc2d3c55f42b8fb79fd.

Some functions to create properties (drm_plane_create_zpos_property or
drm_plane_create_color_properties for example) will ask for a range of
acceptable value and an initial one.

This initial value is then stored in the values array for that property.

Let's provide an helper to access this property.

Acked-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Reviewed-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20220221095918.18763-7-maxime@cerno.tech

6by9 committed:
drm/object: Add default zpos value at reset

Upstream commit 1a7998dab5dd3d11bada7e3921781922082e7fe6

The drm_plane_create_zpos_property() function asks for an initial value,
and will set the associated plane state variable with that value if a
state is present.

However, that function is usually called at a time where there's no
state yet. Since the drm_plane_state reset helper doesn't take care of
reading that value when it's called, it means that in most cases the
initial value will be 0, or the drivers will have to work around it.

Let's add some code in __drm_atomic_helper_plane_state_reset() to get
the initial zpos value if the property has been attached in order to fix
this.

Reviewed-by: Harry Wentland <harry.wentland@amd.com>
Reviewed-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20220221095918.18763-8-maxime@cerno.tech

6by9 committed:
drm/object: Add default color encoding and range value at reset

Upstream commit 9a48ab11714c955456fefdd4ab532d324fbef563

The drm_plane_create_color_properties() function asks for an initial
value for the color encoding and range, and will set the associated
plane state variable with that value if a state is present.

However, that function is usually called at a time where there's no
state yet. Since the drm_plane_state reset helper doesn't take care of
reading that value when it's called, it means that in most cases the
initial value will be 0 (so DRM_COLOR_YCBCR_BT601 and
DRM_COLOR_YCBCR_LIMITED_RANGE, respectively), or the drivers will have
to work around it.

Let's add some code in __drm_atomic_helper_plane_state_reset() to get
the initial encoding and range value if the property has been attached
in order to fix this.

Reviewed-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20220221095918.18763-19-maxime@cerno.tech

6by9 commented:
Rebased and updated to pick up mripard's upstream versions of the patches, plus the patch implementing zpos on vc4.

pelwell commented:
Does the `normalize_zpos` property guarantee that all zpos values are contiguous starting from 0?
